src/shell.html

```diff
@@ -1224,30 +1224,36 @@
       var Module = {
         preRun: [],
         postRun: [],
-        print: (function() {
-          var element = document.getElementById('output');
-          if (element) element.value = ''; // clear browser cache
-          return function(text) {
-            if (arguments.length > 1) text = Array.prototype.slice.call(arguments).join(' ');
-            // These replacements are necessary if you render to raw HTML
-            //text = text.replace(/&/g, "&amp;");
-            //text = text.replace(/</g, "&lt;");
-            //text = text.replace(/>/g, "&gt;");
-            //text = text.replace('\n', '<br>', 'g');
-            console.log(text);
-            if (element) {
-              element.value += text + "\n";
-              element.scrollTop = element.scrollHeight; // focus on bottom
-            }
-          };
-        })(),
-        printErr: function(text) {
-          if (arguments.length > 1) text = Array.prototype.slice.call(arguments).join(' ');
-          if (0) { // XXX disabled for safety typeof dump == 'function') {
-            dump(text + '\n'); // fast, straight to the real console
-          } else {
-            console.error(text);
+        fdBuf: ['', '', ''],
+        printChars: function(strOrOffset, fd, len, buffer) {
```

Comment:
The input parameters and their optionality feels still a bit off. Usually in APIs when there is a (offset, length) pair being handled, the offset and length are usually two consecutive parameters, but here we have a `offset, <unrelated>, length` syntax. I think that we should define this as having the following possible usage forms:

```javascript
// String-based overload:
Module.printChars(1, "Hello stdout!");
Module.printChars(2, "Hello stderr!");
// ArrayBuffer-based overload:
Module.printChars(1, 10/*offset*/, 63/*length*/, HEAPU8); // Prints UTF8 string in HEAPU8[10]...HEAPU8[10+63]
Module.printChars(1, 10, 63); // When source buffer is not specified, then HEAPU8 is used.
Module.printChars(1, 10); // length is also optional, by virtue of how UTF8ArrayToString() is being changed in this PR.
Module.printChars(1, 10, undefined, myArray); // length can be passed as undefined in the middle to specify unbounded read from a custom array.
```

Then the code would read

```javascript
printChars: function(fd, str, len, buffer) {
  if (typeof(str) === 'number') str = UTF8ArrayToString(buffer || HEAPU8, str, len);
  ...
```

and `fd` would never be optional.